### DISCUSSION TITLE
Error importing DXF from AutoCAD

### DISCUSSION THREAD

**Discussion by EngDM** [robbievanleeuwen/section-properties#229] [Q&A] [answered]
Hey all,

First off, from what I've read through the documentation this is going to work fantastically for me should I ever figure out how to run it. Currently I am getting the following error.
```
---------------------------------------------------------------------------
CadImporterError                          Traceback (most recent call last)
<ipython-input-12-4b85f0080413> in <module>
----> 1 geom = Geometry.from_dxf(dxf_filepath='Test.dxf')

I:\Anaconda\lib\site-packages\sectionproperties\pre\geometry.py in from_dxf(dxf_filepath)
    217         :vartype dxf_filepath: Union[str, pathlib.Path]
    218         """
--> 219         return load_dxf(dxf_filepath)
    220 
    221     @classmethod

I:\Anaconda\lib\site-packages\sectionproperties\pre\geometry.py in load_dxf(dxf_filepath)
   1818     my_dxf = c2s.dxf.DxfImporter(dxf_filepath)
   1819     my_dxf.process()
-> 1820     my_dxf.cleanup()
   1821 
   1822     polygons = my_dxf.polygons

I:\Anaconda\Lib\site-packages\cad_to_shapely\cadimporter.py in cleanup(self, simplify, zip_length, retry_with_zip)
     92     def cleanup(self, simplify = True, zip_length = 0.000001, retry_with_zip = True) -> str:
     93         logging.info("cleanup is depreciated BTW. Use the polygonize function instead")
---> 94         self.polygonize(simplify,zip_length,retry_with_zip)
     95         return 'done'
     96 

I:\Anaconda\Lib\site-packages\cad_to_shapely\cadimporter.py in polygonize(self, simplify, force_zip, zip_length, retry_with_zip)
     68 
     69         if not self.geometry:
---> 70             raise CadImporterError('Cannot run polygonize() since no geometry yet. Have you run process()?')
     71 
     72         if not force_zip:

CadImporterError: Cannot run polygonize() since no geometry yet. Have you run process()?
```
I'm trying to use Jupyter due to being a novice, but it appears that it is recognizing the path I just don't know why it won't recognize the geometry in the file. Unfortunately I cannot attach a DXF file, but it is literally a square drawn with linked polylines centered at the origin of the file.

My goal eventually is to be able to draw the current condition beam and then draw seperate shapes for the reinforcement.

EDIT: Below is the error trying to recreate on a different machine.

```
---------------------------------------------------------------------------
CadImporterError                          Traceback (most recent call last)
Input In [4], in <cell line: 1>()
----> 1 geom = Geometry.from_dxf(dxf_filepath="TestPoly.dxf")
      2 geom.plot_geometry()

File ~\anaconda3\lib\site-packages\sectionproperties\pre\geometry.py:219, in Geometry.from_dxf(dxf_filepath)
    209 @staticmethod
    210 def from_dxf(
    211     dxf_filepath: Union[str, pathlib.Path],
    212 ) -> Union[Geometry, CompoundGeometry]:
    213     """

    214     An interface for the creation of Geometry objects from CAD .dxf files.
    215 
    216     :cvar dxf_filepath: A path-like object for the dxf file
    217     :vartype dxf_filepath: Union[str, pathlib.Path]
    218     """
--> 219     return load_dxf(dxf_filepath)

File ~\anaconda3\lib\site-packages\sectionproperties\pre\geometry.py:1820, in load_dxf(dxf_filepath)
   1818 my_dxf = c2s.dxf.DxfImporter(dxf_filepath)
   1819 my_dxf.process()
-> 1820 my_dxf.cleanup()
   1822 polygons = my_dxf.polygons
   1823 new_polygons = c2s.utils.find_holes(polygons)

File ~\anaconda3\Lib\site-packages\cad_to_shapely\cadimporter.py:94, in CadImporter.cleanup(self, simplify, zip_length, retry_with_zip)
     92 def cleanup(self, simplify = True, zip_length = 0.000001, retry_with_zip = True) -> str:    
     93     logging.info("cleanup is depreciated BTW. Use the polygonize function instead")
---> 94     self.polygonize(simplify,zip_length,retry_with_zip)
     95     return 'done'

File ~\anaconda3\Lib\site-packages\cad_to_shapely\cadimporter.py:70, in CadImporter.polygonize(self, simplify, force_zip, zip_length, retry_with_zip)
     67 def polygonize(self, simplify = True, force_zip = False, zip_length = 0.000001, retry_with_zip = True):
     69     if not self.geometry:
---> 70         raise CadImporterError('Cannot run polygonize() since no geometry yet. Have you run process()?')
     72     if not force_zip:
     73         result, dangles, cuts, invalids = ops.polygonize_full(self.geometry)

CadImporterError: Cannot run polygonize() since no geometry yet. Have you run process()?
```
```
geom = Geometry.from_dxf(dxf_filepath="TestRegion.dxf")
geom.plot_geometry()
```
The file is saved in the same folder as the drawing. I've attached a text file.
[TestPoly.txt](https://github.com/robbievanleeuwen/section-properties/files/9002588/TestPoly.txt)

**Comment by jonathanweinberg** [comment]
@EngDM I noticed over in #185 that you might be able to post the .dxf file as a .txt successfully. Best of luck!

**Reply by robbievanleeuwen** [reply]
@EngDM also have a peek at #185 to see if the issue is similar to what you've experienced.

**Comment by robbievanleeuwen** [comment]
Hi @EngDM, as @jonathanweinberg mentioned it would be great if you could upload your ``.dxf`` file (change extension to ``.txt`` should work or host the file somewhere and share the link).

If you could also post your code here that will help with the debugging :)

I'm sure we'll be able to get it working for you!

**Reply by robbievanleeuwen** [reply]
Also a heads up that the pre-processor is now very powerful and you can create and combine many shapes very easily, see [here](https://sectionproperties.readthedocs.io/en/latest/rst/geom_mesh.html), [here](https://sectionproperties.readthedocs.io/en/latest/rst/section_library.html) and [here](https://sectionproperties.readthedocs.io/en/latest/rst/advanced_geom.html).

Importing CAD files also works though!

**Reply by EngDM** [reply]
Unfortunately some of the joists I deal with are weird shapes that are easiest to just draw in AutoCAD instead of finding points.

I'm working to upload my code however I am not on the same system. If anything I will be able to upload it later on today.

**Reply by robbievanleeuwen** [reply]
Thanks for uploading your file @EngDM. I tried it myself and can reproduce your error.

*sectionproperties* uses this great library ([cad-to-shapely](https://github.com/aegis1980/cad-to-shapely)) by @aegis1980, to read dxf files and convert them to shapely geometries. I believe the supported geometries are Splines, Polylines, Lines and Arcs.

I find ``.dxf`` files extremely difficult to parse (there are over 13000 lines in your file for a simple square!). I'm not sure what program you are using the generate the ``.dxf`` file, but if possible I would try playing around with the export options and you might find that a certain profile works? E.g. I've used Rhino in the past and the 2004 LINES AutoCAD export scheme always seemed to work for me.

Failing this, you could always try raising an issue on [cad-to-shapely](https://github.com/aegis1980/cad-to-shapely) as the developer has been super helpful in the past :)

**Reply by EngDM** [reply]
I'm using AutoCAD LT 2022 and saving as a 2018 DXF file. Perhaps I try to use a different save. By profile do you mean just the format to save the file?

Is there anything you can see wrong with my code? I just want to upload a shape and have it calculate the warping properties.

**Reply by robbievanleeuwen** [reply]
I've never used AutoCAD before but if there are any save options for ``.dxf`` files you could try playing around with them. Also wondering if there are different ways to draw the geometry in AutoCAD? Maybe just try using the simplest straight lines as a test (not polylines). Another option maybe is to try to convert your autocad file to a rhino file ``.3dm`` and try to use the rhino import function in *sectionproperties*? I think your best port of call for the ``.dxf`` issue is to raise an issue on cad-to-shapely as this isn't really a *sectionproperties* implementation.

The code should be straightforward:

```python
from sectionproperties.pre.geometry import Geometry

geom = Geometry.from_dxf(dxf_filepath="TestPoly.dxf")
geom.plot_geometry()
```

**Reply by EngDM** [reply]
How were you able to tell the number of nodes in the DXF file?

**Reply by robbievanleeuwen** [reply]
Didn't mean geometric lines, meant there were 13000 lines of text in the ``.dxf`` file. Can't say I'm familiar with the syntax of these files at all.

**Reply by aegis1980** [reply]
@EngDM 
Hey, as @robbievanleeuwen said - its a problem with [cad to shapely](https://pypi.org/project/cad-to-shapely/), which I take care of.
The 13k text lines in DXF file to describe 12-line shape does seem odd, but the dxf file u uploaded seems to load up ok in Rhino. 
Will be something 'weird' and (likely) simple! Give me the weekend ;) ~~what are you generating DXF with?~~ (AutoCAD LT!)
FYI I (re)exported as DXF from Rhino and worked ok....
....FYI2 A while ago I set up a little [Streamlit webapp](https://aegis1980-streamlit-sectionprops-streamlit-app-8lmaoc.streamlitapp.com/) to spew out your section props from imported DXF. Section might look familiar. 

![image](https://user-images.githubusercontent.com/4284988/176362549-d15281b6-2399-4b40-be4a-b956ae8898d8.png)

**Reply by robbievanleeuwen** [reply]
@aegis1980 legend, thanks mate!

Love the streamlit, works great! Mind if I share?

**Reply by EngDM** [reply]
@aegis1980  For exporting from Rhino are you using the software itself or is there a conversion process to a 3dm file that you are aware of? I do not have Rhino.

That Web app is wicked. I'll be using that once we figure out what the issue with my DXF files.

In the future I plan on adding some Markdown code for a more report-like presentation for my engineering records.

@robbievanleeuwen I briefly looked that sectionproperties calculates Mrx and Mry, are these based off any codes or strictly FEA analysis.

**Reply by aegis1980** [reply]
Yeah, rhino is my go-to. import/exports pretty much anything you wish to.

**Reply by aegis1980** [reply]
@robbievanleeuwen share & whatever. it's on streamlit free plan so shuts down and needs to be woken up once in a while, You did all the work! I just stuck it in dumb app to try get people at work excited!
They weren't. Each to their own.

**Reply by robbievanleeuwen** [reply]
@EngDM, if you provide *sectionproperties* geometries with materials, a plastic moment is calculated which assumes all materials are experiencing their yield stress (this is the definition used to calculate the plastic centroid), so there is no material non-linearity or any reference to design codes. If you don't give *sectionproperties* geometries any material properties then you get the plastic section moduli which can be used for design e.g. steel.

For some materials e.g. compact steel sections under certain cases, the plastic moment is what you need for design. However, if you're interested in calculating ultimate concrete capacities this obviously is not very helpful, but watch [this space](https://github.com/robbievanleeuwen/concrete-properties) as there's something coming soon 😉

**Reply by EngDM** [reply]
@robbievanleeuwen @aegis1980 Don't want you guys to think I'm ignoring you, it's a holiday weekend so I will be out of town with limited access to internet. I am looking forward to what aegis discovers over the weekend! Thanks again all.

**Reply by EngDM** [reply]
@robbievanleeuwen Hey Robbie, are you able to upload a DXF with multiple shapes and the code to import it into sectionproperties? For whatever reason my compound geometry isn't recognizing a second shape. I cannot for the life of me figure out what is going on with this.

**Reply by robbievanleeuwen** [reply]
@EngDM are you able to share the dxf file you are having trouble with?

**Reply by EngDM** [reply]
See the attached text file.
[TestHat.txt](https://github.com/robbievanleeuwen/section-properties/files/9164333/TestHat.txt)

**Reply by robbievanleeuwen** [reply]
I've investigated this after installing the ``dev`` branch from ``cad-to-shapely`` to handle the circles.

Looking at [``geometry.py``](https://github.com/robbievanleeuwen/section-properties/blob/master/sectionproperties/pre/geometry.py#L1831), ``cad-to-shapely`` successfully imports all the polygons at line 1831. However after calling the ``cad-to-shapely`` function ``c2s.utils.find_holes(polygons)`` the two circles are removed.

Perhaps this is because the circles are not connected to the top hat section and you aren't 'supposed' to be using *sectionproperties* on unconnected sections (see warning [here](https://sectionproperties.readthedocs.io/en/latest/rst/analysis.html#warping-analysis) - although geometric & plastic analyses work ok for unconnected sections, warping analyses do not).

@aegis1980 was this a design decision for ``cad-to-shapely`` or is there a way to get the ``find_holes()`` function to keep the unconnected geometries?

**Reply by EngDM** [reply]
Thank you for investigating Robbie, when first starting to figure this out a month ago I remember reading that warning but I couldn't for the life of me remember where it was so I started second guessing if that was from this program or something else. One problem I am having is that sectionproperties is not recognizing the holes. What I did was move the circles over to have connectivity and trim the intersections. I will attach a .txt when I am at my personal computer.

[TestHat.txt](https://github.com/robbievanleeuwen/section-properties/files/9172578/TestHat.txt)

Edit: I ran the file again and it appears to recognize the holes, but does not mesh around them. Weird that it sees the holes now.

**Reply by robbievanleeuwen** [reply]
@EngDM,

Seems to mesh fine for me (using the ``dev`` branch of ``cad-to-shapely``). Modified material so you can see what is meshed and what isn't.

```python
from sectionproperties.pre.geometry import Geometry
from sectionproperties.pre.pre import Material
from sectionproperties.analysis.section import Section

mat = Material("test", 1, 0, 1, 1, "grey")

g = Geometry.from_dxf("TestHat.txt")
g.material = mat
g.plot_geometry()
g.create_mesh(10)

sec = Section(g)
sec.plot_mesh()
```
<img width="1031" alt="Screen Shot 2022-07-23 at 4 30 35 pm" src="https://user-images.githubusercontent.com/18841910/180593450-75cea430-de3f-4163-80e4-3b720be93940.png">
<img width="1010" alt="Screen Shot 2022-07-23 at 4 30 46 pm" src="https://user-images.githubusercontent.com/18841910/180593452-6c8cd081-2c0b-4a69-9b2d-5759bc4a4ba0.png">

**Reply by robbievanleeuwen** [reply]
Analysis works as well:
<img width="980" alt="Screen Shot 2022-07-23 at 4 33 37 pm" src="https://user-images.githubusercontent.com/18841910/180593513-c086cc19-37e7-47b5-867d-e36cd2bb59f2.png">

**Reply by EngDM** [reply]
I was running my mesh as [1] and it didn't mesh correctly around the holes. Once I switched to 10 it seems to work. On another note, how do you plot your figures so large? Might be just because I'm using Jupyter Notebook and you're using something else.

**Reply by robbievanleeuwen** [reply]
I can't reproduce this, when passing ``[1]``  to ``create_mesh()`` I get this:

<img width="1076" alt="Screen Shot 2022-07-25 at 12 29 11 pm" src="https://user-images.githubusercontent.com/18841910/180681481-acd0b4f7-3e46-43d2-bae4-a241e060467d.png">

I just run my code through the terminal so can make the plot windows as big as I want. 

You could try passing the ``figsize`` argument to any plotting function in *sectionproperties, e.g.

```python
sec.plot_mesh(figsize=(12, 7))
```

In jupyter, I also find changing the plotting output to ``.svg`` generates nicer vector style plots:

```python
%config InlineBackend.figure_formats = ['svg']
```

**Comment by aegis1980** [comment, marked as answer]
Think cad-to-shapely fixed. v0.3.1 on pypi so ought to show up in `section-properties` soon enough once bots have done their thing.

@EngDM's DXF file was using `LWPOLYGON` entity to describe top-hat geometry. Its a lighter weight version of 'heavy' older POLYGON  and in newer DXF versions. I had not come across it yet and only really implement thing as they come up. 

@EngDM (and anyone else with DXF import issues) if u have (further) problems, trying saving as older DXF version in AutoCAD (or whatever).  But still flag issue up! Best flagged up on on [cad-to-shapely](https://github.com/aegis1980/cad-to-shapely) repo.

Also 13000 lines in DXF file was cos there were hatches embedded in file in raster (tif) format. Not an issue, _per se_, but there might be a 'save small' or similar option to strip this stuff out.

**Reply by robbievanleeuwen** [reply]
@aegis1980 love your work! I can now process the file posted by @EngDM in the original post to work after upgrading ``cad-to-shapely`` with:

```shell
pip install cad_to_shapely --upgrade
```

**Reply by EngDM** [reply]
I've got it working for a simple hat shape, now just to figure out how to get compoundgeometry to recognize multiple shapes. @aegis1980 what would shapely's best option to process a circle be in your opinion? I've tried it with drawing a circle and with drawing a polyline as an arc and joining all ends, neither seem to work.

Edit: I've got one other idea, a dynamic block that basically creates a circle approximation out of many polylines. We use roundbar to reinforce a lot of joists. I'm not sure how to create this, or if shapely will handle blocks well but I will look into it.

**Reply by aegis1980** [reply]
@EngDM Dunno. Not sure I understand question TBH (i do!*). Think this might a separate question. 
Probably didn't get round to adding import CIRCLE to cad-to-shapely (if that is what u are asking?). Can if useful. 
** Allergic to reinforced concrete design since uni

**Reply by robbievanleeuwen** [reply]
@EngDM *sectionproperties* only handles straight lines so circles must be discretised. See how we do it with [circular sections](https://sectionproperties.readthedocs.io/en/latest/rst/section_library.html#circular-section) and [area-maintaing circular sections](https://sectionproperties.readthedocs.io/en/latest/rst/section_library.html#circular-section-by-area). If there was a way for you to easily draw this or similar in AutoCAD that would make for an easy import.

Multiple shapes should work with ``CompoundGeometry.from_dxf()``, let us know if you are having any issues with this.

**Reply by robbievanleeuwen** [reply]
Another way to do it would be to add your circles in using *sectionproperties*, e.g.

```python
from sectionproperties.pre.pre import Material
from sectionproperties.pre.geometry import Geometry
import sectionproperties.pre.library.primitive_sections as sp_ps

# define material properties
concrete = Material(
    name='Concrete', elastic_modulus=30.1e3, poissons_ratio=0.2, yield_strength=32,
    density=2.4e-6, color='lightgrey'
)
steel = Material(
    name='Steel', elastic_modulus=200e3, poissons_ratio=0.3, yield_strength=500,
    density=7.85e-6, color='grey'
)

# import complex concrete geometry
geom = Geometry.from_dxf(dxf_filepath="concrete_section.dxf")
geom.material = concrete

# add five 20 mm reinforcing bars
for idx in range(5):
    # create bar with area of 310 mm^2
    bar = sp_ps.circular_section_by_area(area=310, n=16, material=steel)

    # shift bar to correct location (values chosing as an example)
    bar = bar.shift_section(x_offset=40 + 100 * idx, y_offset=40)

    # create hole in concrete at bar location, then add bar
    geom = (geom - bar) + bar
```

**Reply by EngDM** [reply]
<s>Update: It was working without error on my personal computer, but now at work after trying the same updates to the cad_to_shapely it is returning the same error.</s>

I had to use the following:

```
import sys
!{sys.executable} -m pip install cad_to_shapely
!python -m pip install cad_to_shapely --upgrade
```

I guess my work pc has different versions of python, so when updating in jupyter it has the wrong version of pip as a default.

I've attached a file of what I'd like to be able to analyze for warping. Given that the placement of the reinforcing bar depends highly on the geometry of the shape, and sometimes it's not hat shaped so I couldn't just develop a cure-all.

[Test.txt](https://github.com/robbievanleeuwen/section-properties/files/9040907/Test.txt)

@aegis1980 Yes! In the line properties for the circle it shows up as a "CIRCLE" instead of a polyline, likely this is why it won't accept it as a polyline.

As far as adding multiple shapes I found if I overlap them in CAD first and trim them, and then further join them into one polyline it works. I don't think this fix would work if you need multiple materials, and it won't work with circle or arc shapes until a fix is implemented for that.

**Comment by aegis1980** [comment]
i'll take a look - will be something simple/stupid that I did waaaay back.
